src/GraphQL/GraphQLBuilderExtensions.cs

```diff
@@ -649,7 +649,11 @@ public static IGraphQLBuilder AddDocumentListener<TDocumentListener>(this IGraph
             where TDocumentListener : class, IDocumentExecutionListener
         {
             builder.Services.RegisterAsBoth<IDocumentExecutionListener, TDocumentListener>(serviceLifetime);
-            builder.ConfigureExecutionOptions(options => options.Listeners.Add(options.RequestServices!.GetRequiredService<TDocumentListener>()));
+            builder.ConfigureExecutionOptions(options =>
+            {
+                var requestServices = options.RequestServices ?? throw new MissingRequestServicesException();
```

Comment:
@Shane32 I see a lot of places with such pattern (here and in server project). Let's add simple one line extension method:
```cs
public static IServiceProvider RequestServicesOrThrow(this IProvideRequestServices provider) => provider.RequestServices ?? throw new MissingRequestServicesException();

public interface IProvideRequestServices
{
  IServiceProvider? RequestServices { get; set; }
}
```

Then add `IProvideRequestServices` to `ExecutionOptions` and `IResolveFieldContext`.

Reply:
That is a breaking change.  There are a number of non-breaking changes we could do, but adding `IProvideRequestServices` to `IResolveFieldContext` is not one of them.

Non-breaking changes:
1. Add `RequestServicesOrThrow` extension method for `ExecutionOptions`, `ExecutionContext`, `IExecutionContext`, `ValidationContext`, and `IResolveFieldContext`
2. Add `GetRequiredService` extension method for the above classes, which will throw `RequestServicesMissingException` if `RequestServices` is null
3. Have the first statement within `DocumentExecuter` replace a null `RequestServices` property `NullServiceProvider` (which throws `RequestServicesMissingException` for any call to `GetService)

As a breaking change, `IProvideRequestServices` should be added to the aforementioned classes and interfaces for the next major version.

Then for the next major version, we can simplify the added extension methods, if any.

Other possibilities with the next major version:
- Make `ExecutionOptions.RequestServices` required, with an immediate check within `DocumentExecuter`

Reply:
Option 1 (with notes/todo/issue) about new interface in future.

Reply:
OK

Reply:
#3167